onnxruntime/core/providers/cpu/tensor/col2im.cc

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cpu/tensor/col2im.h"
+#include "core/util/math_cpuonly.h"
+
+namespace onnxruntime {
+
+#define REGISTER_COL2IM_TYPED_KERNEL(OP_TYPE, VERSION, TYPE, KERNEL_CLASS)         \
+  ONNX_CPU_OPERATOR_TYPED_MS_KERNEL(                                               \
+      OP_TYPE,                                                                     \
+      VERSION,                                                                     \
+      TYPE,                                                                        \
+      KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<TYPE>()), \
+      KERNEL_CLASS<TYPE>);
+
+REGISTER_COL2IM_TYPED_KERNEL(Col2Im, 1, float, Col2Im);
+
+template <typename T>
+Status Col2Im<T>::Compute(OpKernelContext* context) const {
+  const auto* col_tensor = context->Input<Tensor>(0);
+  const auto* image_shape = context->Input<Tensor>(1);
+  const auto* kernel_shape = context->Input<Tensor>(2);
+
+  int64_t image_shape_size = 1;
+  int64_t kernel_shape_size = 1;
+  TensorShapeVector adjusted_kernel_shape_dims;
+  for (auto i = 0; i < image_shape->Shape().Size(); ++i) {
+    image_shape_size *= image_shape->Data<int64_t>()[i];
```

Comment:
>>Data<int64_t>()[i]; [](http://example.com/codeflow?start=36&length=20)

Data<>() in the loop produces repeated type checks,
Would it not be better to get a span of data once and apply op[] on it both for image_shape and kernel, here and below.

Reply:
One of the customer requests lately is to give more consideration to parallelizing algorithms.

Reply:
The main goal of this PR was to help PyTorch ONNX converter team to evaluate our proposal for ONNX's Col2Im opset 18 spec with a real implementation. That allowed us to validate the n-dimensional flexibility for Col2Im, which is something that most frameworks do not support. This task has been finished, but I thought this PR could still be used as a scaffolding for a more robust implementation soon.

Because of this, this PR introduced a simple implementation for CPU-only Col2Im contrib op. It does not include CUDA support nor is fully optimized yet.

As I am not working closely to ORT core, all the limitations above could be addressed by a colleague more familiar with ORT core on a follow-up PR that 1) changes domain from contrib to ONNX, 2) adds CUDA support and 3) maybe leverage https://github.com/microsoft/onnxruntime/pull/12476 for parallelism.

What do you think? Another option is to close this PR and let someone from ORT to take it over, but it seems [3p customers](https://github.com/pytorch/pytorch/issues/84408) are looking into using it as soon as possible for research work, that could turn up to be one mire ORT fan :)

Reply:
I think this is too much of the discussion for a couple of calls to DataAsSpan<>() and then use the span variable.